sw/host/ot_certs/src/asn1/builder.rs

```diff
@@ -0,0 +1,151 @@
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+//! This module provides a trait to abstract the generation of an ASN1
+//! document. The particularity of this trait is that the base values
+//! are of type `Value<T>` which can either be literals or variables.
+
+use anyhow::{ensure, Result};
+use num_bigint_dig::BigUint;
+
+use crate::asn1::{Oid, Tag};
+use crate::template::Value;
+
+/// Helper function to add a suffix to a name hint.
+pub fn concat_suffix(name_hint: &Option<String>, suffix: &str) -> Option<String> {
+    name_hint.as_ref().map(|s| format!("{}_{}", s, suffix))
+}
+
+/// Trait for an abstract ASN1 builder.
+///
+/// This builder does not specify what the output is, but only provides
+/// basic functions to build an ASN1 document. A possible implementation
+/// would be to output bytes to produce an actual DER encoding for example.
+/// This can also be used to generate C code that calls into a C library to
+/// produce a certificate. For this reason, this builder does not provide any
+/// way to access the result, as this is implementation specific.
+pub trait Builder {
+    /// Push a byte into the ASN1 output.
+    fn push_byte(&mut self, val: u8) -> Result<()>;
+
+    /// Push a tagged boolean into the ASN1 output.
+    fn push_boolean(&mut self, val: bool) -> Result<()>;
+
+    /// Push a tagged integer into the ASN1 output. The name hint can be used by
+    /// the implementation for documentation purpose, or completely ignored.
+    fn push_integer(
+        &mut self,
+        name_hint: Option<String>,
+        tag: &Tag,
+        val: &Value<BigUint>,
+    ) -> Result<()>;
+
+    /// Push a byte array into the ASN1 output, representing an integer. If the provided
+    /// buffer is smaller than the provided size, it will be padded with zeroes. Note that this
+    /// function does not add a tag to the ASN1 output.
+    fn push_integer_pad(
+        &mut self,
+        name_hint: Option<String>,
+        val: &Value<BigUint>,
+        size: usize,
+    ) -> Result<()>;
+
+    /// Push a byte array of fixed length into the ASN1 output. Note that this function does not add a tag to
+    /// the ASN1 output.
+    fn push_byte_array(&mut self, name_hint: Option<String>, val: &Value<Vec<u8>>) -> Result<()>;
+
+    /// Push a tagged string into the ASN1 output.
+    fn push_string(
+        &mut self,
+        name_hint: Option<String>,
+        str_type: &Tag,
+        val: &Value<String>,
+    ) -> Result<()>;
+
+    /// Push a tagged object identifier into the ASN1 output.
+    fn push_oid(&mut self, oid: &Oid) -> Result<()>;
+
+    /// Push tagged content into the ASN1 output. The closure can use any available function of the builder
+    /// and produces the content of the tagged data.
+    fn push_tag(
+        &mut self,
+        name_hint: Option<String>,
+        tag: &Tag,
+        gen: &dyn Fn(&mut dyn Builder) -> Result<()>,
```

Comment:
Nit (you might not actually want this, which is fine):

```suggestion
        gen: &dyn FnOnce(&mut dyn Builder) -> Result<()>,
```

It's a shame these have to be trait objects and not generics, but if I recall you had a reason for `Builder` being virtual.

Reply:
Generics and closures interact very poorly unfortunately which is why I decided to go for trait objects. If we wanted those method to have generics then the code would need to be refactored around a `Asn1Serializer` trait instead of closures. I think it would be a good idea in the future but it requires careful thinking about the interface so I prefer to avoid it here.

This also explains the `Fn` versus the `FnOnce`: an `FnOnce` consume the closure. But when implementing something like:
```rust
fn push_octet_string(
    &mut self,
    name_hint: Option<String>,
    gen: &dyn FnOnce(&mut dyn Builder) -> Result<()>,
) -> Result<()> {
    self.push_tag(name_hint, &Tag::OctetString, &|builder| gen(builder))
}
```
this would be problematic. Indeed, a call to `gen` would move `gen` since it is a `FnOnce`. But `gen` cannot be moved because it is not sized. Therefore the `Builder` trait would have to be sized. But a sized trait is not object safe so I would have to used generics instead of trait objects.

I am not a rust expert but it seems to me that you can have:
- trait objects and `Fn`, or
- generics and `FnOnce`

But mixing trait objects with `FnOnce` does not really work.

Reply:
What you suggest works but now literally every function that interacts with a builder must become generic for no good reason. Furthermore, it hardcodes the final implementation type into the closure type which breaks the abstraction. Consider the following reduced example:
```rust
trait Builder {
    fn hello(&mut self);
}

struct Impl;

impl Builder for Impl {
    fn hello(&mut self) {
        println!("hello");
    }
}
```
Now imagine that you want to implement a function `hello` that uses `Impl` and takes a closure that works for any `Builder`. Something like:
```rust
// The following does not work?
fn example<T: Builder>(gen: impl FnOnce(&mut T)) {
    let mut builder = Impl;
    gen(&mut builder);
}
```
I don't know how to make it work at least. The only solution I know is:
```rust
fn example(gen: impl FnOnce(&mut Impl)) {
    let mut builder = Impl;
    gen(&mut builder);
}
```
which suddenly "leaks" the type `Impl` in the type of example. I don't know how to fix this without replacing the closure by a trait, like serde's `Serialize` for example.

Reply:
Sorry if I'm misunderstanding, but your above example seems fine and doesn't need to be generic since `gen(...)` is only called with `Impl` as its argument type.

What do you mean about leaking the `Impl` type?